File: apps/java/src/main/java/chapter03/DecisionTree.java
package chapter03;

import chapter02.Features;
import java.util.List;
import java.util.Optional;

/** 自作の決定木の分類器。fit で学習してから predict で予測する。 */
public final class DecisionTree {
  private static final int UNLIMITED = -1;

  private final int maxDepth;
  private Tree tree;

  private DecisionTree(int maxDepth) {
    this.maxDepth = maxDepth;
  }

  /** 深さを制限しない決定木。 */
  public static DecisionTree unlimited() {
    return new DecisionTree(UNLIMITED);
  }

  /** 深さの上限を指定した決定木。 */
  public static DecisionTree withMaxDepth(int maxDepth) {
    if (maxDepth < 0) {
      throw new IllegalArgumentException("深さの上限は 0 以上にしてください");
    }
    return new DecisionTree(maxDepth);
  }

  /** 訓練データから木を作る。 */
  public DecisionTree fit(List<Features> x, List<String> t) {
    tree = DecisionTrees.build(x, t, maxDepth);
    return this;
  }

  /** 学習した木。学習する前は空。 */
  public Optional<Tree> tree() {
    return Optional.ofNullable(tree);
  }

  /** 特徴量ごとのラベルを予測する。 */
  public List<String> predict(List<Features> x) {
    if (tree == null) {
      throw new IllegalStateException("fit で学習してから predict を呼んでください");
    }
    return x.stream().map(features -> DecisionTrees.predictOne(tree, features)).toList();
  }
}

File: apps/csharp/src/MachineLearning/Chapter03/DecisionTree.cs
namespace MachineLearning.Chapter03;

using MachineLearning.Chapter02;

/// <summary>自作の決定木の分類器。Fit で学習してから Predict で予測する。</summary>
public sealed class DecisionTree
{
    private const int UnlimitedDepth = -1;

    private readonly int maxDepth;

    private DecisionTree(int maxDepth) => this.maxDepth = maxDepth;

    /// <summary>学習した木。学習する前は null。</summary>
    public Tree? Tree { get; private set; }

    /// <summary>深さを制限しない決定木。</summary>
    public static DecisionTree Unlimited() => new(UnlimitedDepth);

    /// <summary>深さの上限を指定した決定木。</summary>
    public static DecisionTree WithMaxDepth(int maxDepth) =>
        maxDepth < 0
            ? throw new ArgumentOutOfRangeException(nameof(maxDepth), "深さの上限は 0 以上にしてください")
            : new DecisionTree(maxDepth);

    /// <summary>訓練データから木を作る。</summary>
    public DecisionTree Fit(IReadOnlyList<Features> x, IReadOnlyList<string> t)
    {
        this.Tree = DecisionTrees.Build(x, t, this.maxDepth);
        return this;
    }

    /// <summary>特徴量ごとのラベルを予測する。</summary>
    public IReadOnlyList<string> Predict(IReadOnlyList<Features> x)
    {
        ArgumentNullException.ThrowIfNull(x);
        var tree = this.Tree ?? throw new InvalidOperationException("Fit で学習してから Predict を呼んでください");
        return [.. x.Select(features => DecisionTrees.PredictOne(tree, features))];
    }
}
